src/Domain/Services/GoogleAnalytics.php

```diff
@@ -0,0 +1,197 @@
+<?php
+namespace Automattic\WooCommerce\Blocks\Domain\Services;
+
+use Automattic\WooCommerce\Blocks\Package;
+use Automattic\WooCommerce\Blocks\Assets\Api as AssetApi;
+
+/**
+ * Service class to integrate Blocks with the Google Analytics extension,
+ */
+class GoogleAnalytics {
+
+	/**
+	 * Stores a record of inline scripts already enqueued.
+	 *
+	 * @var array
+	 */
+	private $inline_scripts = [];
+
+	/**
+	 * Constructor
+	 */
+	public function __construct() {
+		if ( ! Package::feature()->is_experimental_build() ) {
+			return;
+		}
+		add_action( 'init', array( $this, 'register_assets' ) );
+		add_action( 'init', array( $this, 'inline_block_tracking' ) );
+		add_action( 'wp_enqueue_scripts', array( $this, 'enqueue_scripts' ) );
+		add_filter( 'script_loader_tag', array( $this, 'async_script_loader_tags' ), 10, 3 );
+	}
+
+	/**
+	 * Inline tracking events for blocks which are server side rendered, or saved to the post.
+	 */
+	public function inline_block_tracking() {
+		add_filter(
+			'experimental__woocommerce_blocks_product_grid_render_products',
+			function( $content, $block_name, $products, $attributes ) {
+				$product_impression_objects = array_map(
+					function( $product_id ) {
+						$product = wc_get_product( $product_id );
+						$cats    = get_the_terms( $product->get_id(), 'product_category' );
+						return (object) [
+							'id'        => $product->get_sku() ? $product->get_sku() : '#' . $product->get_id(),
+							'name'      => $product->get_title(),
+							'list_name' => $block_name,
+							'category'  => $cats && ! is_wp_error( $cats ) ? current( $cats )->term_name : '',
+							'price'     => $product->get_price(),
+						];
+					},
+					$products
+				);
+				$this->add_inline_script(
+					'wc-blocks-google-analytics',
+					"
+					gtag(
+						'event',
+						'view_item_list',
+						{
+							event_category: 'engagement',
+                            event_label: '" . esc_js(
+						__(
+							'Viewing products',
+							'woo-gutenberg-products-block'
+						)
+					) . "',
+							items: JSON.parse( decodeURIComponent( '" . rawurlencode( wp_json_encode( $product_impression_objects ) ) . "' ) ),
+						}
+					);
+                    "
+				);
+				return $content;
+			},
+			10,
+			4
+		);
+		add_action(
+			'experimental__product-search_render_callback',
+			function() {
+				$this->add_inline_script(
+					'wc-blocks-google-analytics',
+					"
+					const forms = document.querySelectorAll( '.wc-block-product-search form' );
+
+					for ( const form of forms ) {
+						form.addEventListener( 'submit', () => {
+							const field = form.querySelector( '.wc-block-product-search__field' );
+
+							if ( field && field.value ) {
+								wp.hooks.doAction( 'experimental__woocommerce_blocks-product-search', { searchTerm: field.value } );
+							}
```

Comment:
So this hook is actually being set by GoogleAnalytics and not blocks code? Then the GA script is going to fire on this hook point?

I would expect that instead of this, the rendered blocks would have the hook point that _any_ script could then hook into.

The concern here is this seems too tightly coupled to GoogleAnalytics. Where would this code live when GA specific stuff is moved to the extension?

Reply:
Some of the hooks/points seem pretty unique to GA which is why I was trying to generalise the hooks and have GA handle most of the more niche stuff like this one. I can't think of much uses besides GA for tracking this form submission ¯\_(ツ)_/¯